### DISCUSSION TITLE
How to tell Doxygen to process an #include within a class definition

### DISCUSSION THREAD

**Discussion by itavero** [doxygen/doxygen#10997] [Q&A] [answered]
In our code base, we generated several base classes as part of our build process, which we inherit from to implement application logic (the generated classes take care of the "infrastructure").
In order not to repeat our selves to often, the tool also generates `.inc` file, which looks something like:
```cpp
/**
 * @file some_component_gen.inc
 */

// ***** THIS FILE IS GENERATED. DO NOT MODIFY! *****

public:
  void OnComponentsLoaded() final;
  void OnTimerElapsed() final;
```

This basically lists all the function we need to override from the generated base class, so we do not have to repeat that again in our class that inherits from it. Instead, we include this `.inc` file within the class declaration:
```cpp
/* .. there's more in this file, but just to give you an idea about the usage .. */
namespace components {
class SomeComponent : public BaseSomeComponent {
private:
  void SomethingInternal();

public:
  /* constructor and stuff here */

/* lastly we include the .inc file */
#include "some_component_gen.inc"
};
} // namespace components
```

This works just fine for us, but now we are trying to generate documentation and I see a lot of warnings that look like:
```
warning: no uniquely matching class member found for 
void component::SomeComponent::OnComponentsLoaded()
Possible candidates:
  void component::BaseSomeComponent::OnComponentsLoaded()
  void component::BaseSomeOtherComponent::OnComponentsLoaded()
  void component::BaseYetAnotherComponent::OnComponentsLoaded()
```

As input in the configuration file (I use CMake to generate it via the `doxygen_add_docs` function it offers), I have a combination of the following:
* Include directories
* File paths to source files
* File paths to header files
* File paths to these `.inc` files.

I also added an `EXTENSION_MAPPING = .inc=C++`, after I learned that Doxygen treats `.inc` files as PHP by default.
I've attached a [redacted version of the Doxygen configuration file](https://github.com/user-attachments/files/16190869/Doxyfile_redacted.txt) as a reference.

My question of course is, how do I have Doxygen parse these `.inc` files correctly in the location they are `#include`d?

**Comment by albert-github** [comment]
Difficult problem as doxygen is not a compiler and does only a limited amount of preprocessing.
For the documentation parts there are some features (like: `\include...` and `\snippet...`) but for the "bare" code they are not really there (only e.g. `#define`s are filtered).
Only possibilitiesy I see:
- is to include these line by means of the `FILTER_PATTERNS` etc. into doxygen. The later will break the line number counting and consequently the line numbers in e.g. warnings (so some internal commands have to be used as well)
- rewrite the `.inc` file to e.g.
   ```
   /**
   * @file some_component_gen.inc
   */

  // ***** THIS FILE IS GENERATED. DO NOT MODIFY! *****

  #define some_component_gen \
  public: \
    void OnComponentsLoaded() final; \
    void OnTimerElapsed() final;
   ```
  and "change" the  include statement from:
  ```
  #include "some_component_gen.inc"
  ```
  to
  ```
  #include "some_component_gen.inc"
  some_component_gen
  ```

**Reply by itavero** [reply]
For now I used the first workaround you suggested, but if the regression that occurred in version 1.8.16 can be resolved, that would be even better of course.

**Comment by doxygen** [comment]
@albert-github This actually used to work (try for instance Doxygen version 1.8.9), where any `#include` inside a `{...}` was treated specially and did an actually include just like the preprocessor would. Somewhere between 1.8.9 and 1.8.20 this feature got lost it seems...

**Comment by albert-github** [comment]
Looks like the problem is introduced between versions 1.8.15 and 1.8.16 and to be more precise (checked with the executables I had lying around):
- June 2, 2019 7403c9e4f7f05e4660ed6d3a54f3f935958b01fd  is OK
- June 10, 2019 572035ffcd0e4f7313bab23d375d63008ebbe5a3  is WRONG

So I suspect the change (047df30104f628bdfbbe71fbc8d11440a5b0494a ):
```
Commit: 047df30104f628bdfbbe71fbc8d11440a5b0494a [047df30]
Date: Sunday, June 2, 2019 1:46:31 PM

issue #7021 Warning about missing item in INCLUDE_PATH

Do the `curlyCount` per file and not globally.
```
to be responsible.
Might be that we need some special handling for `extern "C" {`

A related issue might be: https://github.com/doxygen/doxygen/issues/10272

**Reply by itavero** [reply]
That issue indeed shows the same problem as far as I can tell.

**Comment by doxygen** [comment, marked as answer]
@itavero I've just pushed a fix for this, see commit https://github.com/doxygen/doxygen/commit/6ae0631c9792a343e21596b4a18c22afb88248dd. Let me know if this works for you.

**Reply by albert-github** [reply]
@doxygen in the minimal example of #7021 I now get again the warning:
```
warning: include file stdio.h not found, perhaps you forgot to add its directory to INCLUDE_PATH?
```
although it is included as `#include <stdio.h>`

Example: [example.tar.gz](https://github.com/user-attachments/files/16194934/example.tar.gz)

**Reply by doxygen** [reply]
Not sure what to do about this, other than remove the warning...

**Reply by doxygen** [reply]
Ok, one nuance I can think of, is to give the warning only for local includes (`"..."`), not for system includes (`<...>`). Implemented that with commit https://github.com/doxygen/doxygen/commit/9c6b11330ef249d715ef84fa8a968c6aaadd5330

**Reply by albert-github** [reply]
I don't think / I'm not sure that removing the warning is a good idea.
In the first occurrence of the `#include <stdio.h>` in `Include.h` no warning is given.
We also see that the warning with the original problem reappears (which would suggest that removing the warning is the best possibility.

Maybe an approach with includes that use `<..>` the warning is not given (normally signalling system files) and thus only when using `"` the warning is given.
(I see that this is now implemented, I was looking into the issue and missed the update).

**Reply by doxygen** [reply]
Indeed. The example is also very strange, including a system include into a user defined namespace is asking for trouble.

**Reply by albert-github** [reply]
I agree that the example is strange, but users ...

**Reply by itavero** [reply]
> Let me know if this works for you.

Won't have time to verify it today, but I'll try to do so next week. Are there build artifacts that can be downloaded somewhere, or do I have to compile it myself?

**Reply by albert-github** [reply]
See the GitHub actions: https://github.com/doxygen/doxygen/actions select the last master build and on the resulting page you will find some artifacts.

**Reply by itavero** [reply]
Just checked the latest `master` build (`1.12.0 (9152d391dc98515f38f2dfb32ce4f45b4f060d90*)`) and this works for me without the additional filter.
I no longer get the warnings mentioned and it seems to display the inherited methods correctly as well.

As you can imagine, I'd love to know when a next release is expected, so I can put a note on our backlog to update our CI image and configuration management scripts to update to this release in our company as well.

**Reply by itavero** [reply]
@albert-github I'm afraid a regression occurred as I see the warnings again on 1.13.2 (26342b775ea25e6fefb53220926b20702c56fcb3) and 1.14.0 (cbe58f6237b2238c9af7f51c6b7afb8bbf52c866). I haven't tested other versions yet to pin point when they got back and I'm still researching if anything else has changed. Just wanted to give a heads up.

**Reply by albert-github** [reply]
I tested with the version 1.11.0 .. 1.16.1 and I didn't see a problem (maybe I missed something).
With doxygen 1.11.0 ( so the faulty one) I see on the functions.html page:
<img width="581" height="157" alt="image" src="https://github.com/user-attachments/assets/be7ebe86-b769-40c6-893e-7a859a6e607c" />

with all other, newer, versions I see:
<img width="474" height="127" alt="image" src="https://github.com/user-attachments/assets/b5f6bdf1-ce5b-4dea-9fc8-a4ac86621595" />


Example used: [example.tar.gz](https://github.com/user-attachments/files/25297015/example.tar.gz)

**Reply by itavero** [reply]
I just tried v1.16.1 as well and I also see the issue there. Let me check if something in our build tooling changed that may have caused this.

**Reply by itavero** [reply]
Interestingly, with 1.12.0 (c73f5d30f9e8b1df5ba15a1d064ff2067cbb8267) it also doesn't work.

In the mean time, with the help of an LLM, I also went through the history of the relevant sources in our code base, but so far the few differences I did see did not influence the results (I tried reverting those changes and running Doxygen again, yielding the same warnings).

Of course, at the time I tested with the build of 9152d391dc98515f38f2dfb32ce4f45b4f060d90 so perhaps something else changed.
I'm not sure if I ever did a retest with the actual release when integrating it (I'm starting to doubt that now).

**Reply by albert-github** [reply]
Did you try the example I attached with a plain doxygen version i.e. just stating doxygen ?

**Reply by itavero** [reply]
I tried the example which worked correctly, and after a lot of bisecting our repositories history starting from July 2024, I finally found that there was an faulty change introduced in our Doxygen configuration.
Somehow a pattern ended up in `EXCLUDE_PATTERNS` that caused our `.inc` files to be excluded.

So sorry to have bothered you for this. I am really thankful for your quick response though.

**Reply by albert-github** [reply]
@itavero 
Thanks, most important is that the problem is solved
